Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1540,17 +1540,46 @@ extension BrowserViewController: URLBarDelegate {
     }
 
     func urlBar(urlBar: URLBarView, didSubmitText text: String) {
-        // If we can't make a valid URL, do a search query.
-        // If we still don't have a valid URL, something is broken. Give up.
-        let engine = profile.searchEngines.defaultEngine
-        guard let url = URIFixup.getURL(text) ??
-                        engine.searchURLForQuery(text) else {
-            log.error("Error handling URL entry: \"\(text)\".")
+        let trimmedText = text.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet())
+        
+        guard trimmedText.contains(" ") else {
+            submitText(text)
             return
         }
 
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ")!
+        let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+        let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+        
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) {
+                if let urlString = $0.successValue {
+                    let engine = self.profile.searchEngines.defaultEngine
 
+                    guard let url = engine.searchURLForQuery(possibleQuery, searchTemplate: urlString) else {
```

Comment:
I think the function you're talking about (`getURLFromTemplate`) does way more than we need -- it should just be a simple encoding of the search terms. Something like this should be enough:

`let escapedQuery = query.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.SearchTermsAllowedCharacterSet())`

(`SearchTermsAllowedCharacterSet` should be moved to `NSCharacterSetExtensions.swift`).

Keyword searching isn't related to search engines, so we should probably avoid using it here.